### PR TITLE
kqueue: use EVFILT_USER as wakeup event if available

### PR DESCRIPTION
**Pull request by panjf2000**
Eliminate the kernel overhead of `pipe` if we can.

### CODE REVIEW

**Comment by panjf2000** [comment]
Huh, a few confusing test failures surface.

**Comment by bnoordhuis** [comment]
I experimented with EVFILT_USER around 2013 and I remember it was broken on macos. Looks like it still is?

I'm afraid I don't really remember the details. Possibly it only worked as a oneshot event, although it seems libevent uses it in multishot mode.

**Comment by panjf2000** [comment]
> I experimented with EVFILT_USER around 2013 and I remember it was broken on macos. Looks like it still is?
> 
> I'm afraid I don't really remember the details. Possibly it only worked as a oneshot event, although it seems libevent uses it in multishot mode.

`EVFILT_USER` doesn't seem to exist in the [macOS man pages](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/kqueue.2.html) anymore. Maybe there is indeed something wrong with `EVFILT_USER`.

**Comment by panjf2000** [comment]
Besides, it looks like `EVFILT_USER` is available on NetBSD since this [commit](https://mail-index.netbsd.org/netbsd-bugs/2020/09/16/msg069221.html).

Unfortunately, FreeBSD and NetBSD are not available with GitHub Actions, but something like [cross-platform-actions](https://github.com/cross-platform-actions/action) may come to help.

File: src/unix/async.c
Comment (bnoordhuis) [inline src/unix/async.c:143]: This should probably be defined somewhere central (e.g. in src/unix/internal.h) because right now the ifdef guards are a mix of:

- if EVFILT_USER
- if EVFILT_USER && NOTE_TRIGGER
- if EVFILT_USER && NOTE_TRIGGER && !__APPLE__

That doesn't look right to me. It's skipping the read loop now but doesn't call kevent() on macos.

Comment (panjf2000) [reply src/unix/async.c:143]: > That doesn't look right to me. It's skipping the read loop now but doesn't call kevent() on macos.

Actually, we only include <event.h> that defines `EVFILT_USER` and `NOTE_TRIGGER` on freebsd and netbsd (see the top of this file), so this preprocesser directive will fail and go into the read loop on macOS.

Comment (panjf2000) [reply src/unix/async.c:143]: > This should probably be defined somewhere central

But I'm with you on that, working on it.

Comment (panjf2000) [reply src/unix/async.c]: Done.

**Comment by vtjnash** [comment]
Would it be sensible to try creating a zero-second one shot timer timeout instead?

**Comment by panjf2000** [comment]
> Would it be sensible to try creating a zero-second one shot timer timeout instead?

I don't think this is the right path. In that way, we're unable to ensure a real-time interruption to `kqueue/epoll/...` cuz it needs to wait for the in-flight `uv__io_poll()` to finish or for the next round to begin.

**Comment by vtjnash** [comment]
Isn't it waiting for those anyways?

**Comment by panjf2000** [comment]
> Isn't it waiting for those anyways?

No, it isn't. And I think that's why `async.c` is created, to allow other threads to emit real-time interruptions to the main thread, waking it up from the polling immediately. Please correct me if I'm missing something.

**Comment by vtjnash** [comment]
A zero second timer should wake it immediately too, though I won't claim to know if that is a useful alternative to the existing pipe

**Comment by panjf2000** [comment]
Kindly ping @bnoordhuis

**Comment by bnoordhuis** [comment]
After thinking it over, I'm somewhat disinclined to merge this. It's for niche platforms that I personally don't care about, the changes get intermingled with platform code that I do care about, and the real-world performance improvements aren't clear.

I could maybe be persuaded if it gave a really good bump on the async handles benchmarks (grep test/benchmark-list.h for "async") but no promises.

**Comment by panjf2000** [comment]
> After thinking it over, I'm somewhat disinclined to merge this. It's for niche platforms that I personally don't care about, the changes get intermingled with platform code that I do care about, and the real-world performance improvements aren't clear.
> 
> I could maybe be persuaded if it gave a really good bump on the async handles benchmarks (grep test/benchmark-list.h for "async") but no promises.

From where I stand, well-testing can justify merging code of improvement. Besides, NetBSD may not catch your attention, but FreeBSD is the Tier-2 support of `libuv` (also the most popular *BSD system). So if we can run the test suite of `libuv` on FreeBSD and no errors occur, I'd say it is worth going along.

**Comment by bnoordhuis** [comment]
Let's see where other maintainers stand on this.

**Comment by panjf2000** [comment]
I've finally discovered the root cause that had been failing the tests on macOS before, `uv__async_send()` might get called before `uv__io_poll()` with multi-threads, which resulted in an `ENOENT` returned from `kevent()` and the process got aborted.

**Comment by panjf2000** [comment]
There might be some code conflicts with #4400.

File: src/unix/async.c
Comment (vtjnash) [inline src/unix/async.c:191]: Out of curiosity, would it be any simpler / faster to register it as an EVFILT_TIMER that is available everywhere, like so? (with corresponding adjustment to `ev->filter` cases to handle it)

```suggestion
  EV_SET(&ev, loop->async_io_watcher.fd, EV_ADD | EVFILT_TIMER | EV_ONESHOT, 0, 0, 0, 0);
  err = kevent(loop->backend_fd, &ev, 1, NULL, 0, NULL);
```

Comment (panjf2000) [reply src/unix/async.c:191]: I have no idea. AFAIK those renowned cross-platform network frameworks rarely use `EVFILT_TIMER` as the notification mechanism for `kqueue`, they usually use `EVFILT_USER` which is also recommended to be used as a wakeup event. Besides, if `EV_ONESHOT` is set, we have to rearm that event after we consume it, it doesn't look like a pragmatic event for the notification mechanism for `kqueue`.

**Comment by panjf2000** [comment]
Some references of `EVFILT_USER` on various platforms:

- [FreeBSD kqueue](https://man.freebsd.org/cgi/man.cgi?query=kqueue)
- [NetBSD kqueue](https://man.netbsd.org/NetBSD-10.0/kqueue.2)
- [DragonFlyBSD annotated tag v4.0.0rc created](https://lists.dragonflybsd.org/pipermail/commits/2014-October/343769.html)
- [macOS nio-swift](https://github.com/apple/swift-nio/blob/196ceab4b21ba5b84eea522f9829db31b7b91f19/Sources/NIOPosix/SelectorKqueue.swift#L167-L182)
- [XNU kernel](https://github.com/apple-oss-distributions/xnu/blob/94d3b452840153a99b38a3a9659680b2a006908e/tests/kqueue_in_select.c#L42-L50)
-

**Comment by vtjnash** [comment]
Okay, it wasn't clear that EVFILT_USER was available on XNU. Makes more sense now that is clear it is available in the XNU sources, just not documented. Could you run the benchmark on this?

**Comment by panjf2000** [comment]
> Okay, it wasn't clear that EVFILT_USER was available on XNU. Makes more sense now that is clear it is available in the XNU sources, just not documented. Could you run the benchmark on this?

### Environment
     Model : Mac Studio (2022)
        OS : macOS 14.1.2
       CPU : Apple M1 Max, 10-core CPU with 8 performance cores and 2 efficiency cores
    Memory : 32GB unified memory

### Benchmark command
```bash
build/uv_run_benchmarks_a million_async
```

### Benchmark results

Branch: libuv:v1.x
```bash
ok 1 - million_async
# 10,348,986 async events in 5.0 seconds (2,069,797/s, 1,048,576 unique handles seen)
```

Branch: panjf2000:kqueue-wake
```bash
ok 1 - million_async
# 12,977,383 async events in 5.0 seconds (2,595,476/s, 1,048,576 unique handles seen)
```

@vtjnash

**Comment by saghul** [comment]
Nice!

**Comment by panjf2000** [comment]
Any new thoughts on this? @bnoordhuis @vtjnash @saghul

**Review by saghul — APPROVED** [review]

**Comment by panjf2000** [comment]
Ping @libuv/collaborators

**Comment by panjf2000** [comment]
I see @saghul gave his approval on this a long time ago but I guess that will not suffice to merge this PR. Since @bnoordhuis @vtjnash seemed to have got a lot of other work on their plates recently and have no time to make the review here, I'm wondering if there are other available collaborators @libuv/collaborators who could chime in and review this PR?

**Comment by panjf2000** [comment]
> I see @saghul gave his approval on this a long time ago but I guess that will not suffice to merge this PR. Since @bnoordhuis @vtjnash seemed to have got a lot of other work on their plates recently and have no time to make the review here, I'm wondering if there are other available collaborators @libuv/collaborators who could chime in and review this PR?

Another two weeks have passed and yet I still haven't heard back from @libuv/collaborators. I wonder what's going on, could anyone help make progress on this PR? It's been dangling around for a long time. @bnoordhuis @vtjnash @saghul

**Comment by panjf2000** [comment]
Another two weeks have passed, still zero response...

Helplessly and despairingly ping @bnoordhuis @vtjnash @saghul @libuv/collaborators

**Comment by bnoordhuis** [comment]
My thinking here is broadly the same as https://github.com/libuv/libuv/pull/4400#issuecomment-2240044278 (tl;dr debatable cost/benefit ratio.)

I don't want to dismiss your pull requests out of hand but I'm also somewhat reticent to merge them out of pure, unadulterated conservatism. But don't get me wrong, I'm not blocking this outright and maybe other @libuv/collaborators feel differently.

**Review by vtjnash — APPROVED** [review]

File: src/unix/internal.h
Comment (vtjnash) [inline src/unix/internal.h:503]: Nit: libuv normally does not allow compile-time feature detection, as it tends to be unreliable for users. Most of those are EOL, except NetBSD 10.0 which is just released.

Could you change this to add runtime detection (of whether EVFILT_USER succeeded)?

File: src/unix/async.c
Comment (vtjnash) [inline src/unix/async.c:249]: If I understand the docs correctly, you don't need to allocate anything here, so you can instead make this an invalid fd, and also then to use this value later as a sentinel to see whether it can use EVFILT_USER and whether this needs to be closed during cleanup.
```suggestion
  /* This is not used as a fd, just an ident(ifier). */
  pipefd[0] = -2;
```

Comment (panjf2000) [reply src/unix/async.c:249]: Yes, it was supposed to be that way. I planned on using a magic number for this but something like https://github.com/libuv/libuv/blob/5d1ccc12c48099d720bb39f7430c480a52953039/src/unix/core.c#L895 and https://github.com/libuv/libuv/blob/5d1ccc12c48099d720bb39f7430c480a52953039/src/unix/core.c#L908 seemed to leave me no choice unless I'm permitted to make some code changes to those portion of the current code.

File: src/unix/internal.h
Comment (panjf2000) [reply src/unix/internal.h:503]: AFAIK, it seems to me that we've broken that rule many times, cuz you know, there are macros detecting the presence of system calls everywhere in `libuv`.

But your suggestion also sounds rational to me, I think it's doable. Are you leaning towards me utilizing the function like `check_c_source_runs` of `cmake`, or just writing some testing code inside the `uv__kqueue_init()` in `kqueue.c`?

File: src/unix/async.c
Comment (vtjnash) [reply src/unix/async.c:249]: Yeah, you don't necessarily need to go through the generic code for that, since it is only valid in kevent. Potentially similar to how libuv now deals with process exit events there also

File: src/unix/internal.h
Comment (vtjnash) [reply src/unix/internal.h:503]: Is there? I thought it was usually detecting systems not headers, and then doing runtime detection of ENOSYS. Irregardless, it should not be a cmake check, that would be even worse than this current check

File: src/unix/kqueue.c
Comment (vtjnash) [inline src/unix/kqueue.c:84]: We cannot simply `abort` in a condition we know will happen (even on platforms that are quite minor--e.g. NetBSD 9.x). This needs to fall-through to using the previous, non-EVFILT_USER based, approach

File: src/unix/kqueue.c
Comment (vtjnash) [inline src/unix/kqueue.c:77]: Why does it need to trigger the event? Wouldn't the first EV_ADD already fail with EINVAL?

File: src/unix/kqueue.c
Comment (vtjnash) [inline src/unix/kqueue.c:339]: Instead of needing to make these conditional, I'd recommend sticking a `maybe_resize(loop, 30)` call during loop initialization. It is a bit silly to need to grow this several times when adding the first couple fd anyways

**Comment by saghul** [comment]
Sorry for the delay, life got busy...

I share the sentiment with Ben. Won't block the PR outright, but IMHO the complexity this adds for the benefit in a corner case is not an insta-merge.

**Comment by panjf2000** [comment]
I have to admit that the latest commit has bloated this PR. The original version would be way more concise, maybe we should go back to using a placeholder file descriptor for the `EVFILT_USER` event and only keep the runtime detection.

**Comment by panjf2000** [comment]
Hi @vtjnash, do you have a minute to review the latest commit?

File: src/unix/kqueue.c
Comment (vtjnash) [reply src/unix/kqueue.c:79]: This is still true. You need to do runtime detection and fallback to the old implementation, not abort

**Comment by saghul** [comment]
@panjf2000 The current approach doesn't do runtime detection, it simply aborts. That is not something we can land.

**Comment by panjf2000** [comment]
Like I said before, mixing compile-time macro and runtime detection will make the code really messy. Do we really need the runtime detection?

**Comment by saghul** [comment]
We cannot abort in this case because there are legit systems where it will break.

Can't we do the detection with sime uv_once and a static variable?

**Comment by panjf2000** [comment]
I was thinking of `loop->flags` or `loop->flags` or `loop->internal_fields`, but I was not sure if you would like that idea?

**Comment by saghul** [comment]
That could also work indeed.

**Comment by panjf2000** [comment]
I went with `uv_once()` cuz this approach can prevent multiple runtime detections from `fork(2)`. PTAL @vtjnash @saghul

**Comment by panjf2000** [comment]
I went with `uv_once()` cuz this approach can prevent multiple runtime detections from `fork(2)`. PTAL @vtjnash @saghul

**Comment by panjf2000** [comment]
Hah, seems like something irrelevant went wrong on Windows' sanitizer checks?

**Comment by panjf2000** [comment]
> Hah, seems like something irrelevant went wrong on Windows' sanitizer checks?

I believe this was introduced by #4254.

File: src/unix/internal.h
Comment (vtjnash) [inline src/unix/internal.h:516]: looks unused now?

File: src/unix/kqueue.c
Comment (vtjnash) [inline src/unix/kqueue.c:344]: style nit: fix indent
```suggestion
        w = &loop->async_io_watcher;
        assert(fd == w->fd);
        uv__metrics_update_idle_time(loop);
        w->cb(loop, w, w->events);
        nevents++;
        continue;
```

**Comment by vtjnash** [comment]
> I went with uv_once() cuz this approach can prevent multiple runtime detections from fork(2)

Using `fork` in a program linked against pthreads or that uses `pthreads_once` is not supported and not a legal program, so that doesn't really matter. The extra `kqueue` with this approach seems like a bit of extra complexity, but not enough to be an issue, so SGTM.

File: src/unix/async.c
Comment (vtjnash) [inline src/unix/async.c:49]: why is this a 1 second sleep?
```suggestion
  struct timespec timeout = {0, 0};
```

File: src/unix/async.c
Comment (vtjnash) [inline src/unix/async.c:60]: Is all this necessary, or would it suffice just to check if the filter is valid (or if it causes EINVAL)
```suggestion
  EV_SET(ev, 0, EVFILT_USER, EV_ADD | EV_CLEAR, 0, 0, 0);
  if (kevent(kq, ev, 1, NULL, 0, NULL) != 0)
```

**Comment by santigimeno** [comment]
Is this ready to land? Once this lands I'll proceed with the release.

File: src/unix/internal.h
Comment (saghul) [reply src/unix/internal.h:516]: Still used AFAIS.

**Review by saghul — APPROVED** [review]
LGTM, let's wait for @vtjnash to give it one last look too.

**Comment by panjf2000** [comment]
Just rebased the code from the mainline, nothing changed in this PR.

**Comment by vtjnash** [comment]
There appears to be a 1 second sleep in the code at startup that should be removed, but everything else looked good to me

**Comment by panjf2000** [comment]
> There appears to be a 1 second sleep in the code at startup that should be removed, but everything else looked good to me

There will not be a sleep because we've added the wakeup event of EVFILT_USER before start polling. Unless you're talking about those platforms where EVFILT_USERs are broken?

**Comment by vtjnash** [comment]
Yes, that would be still be a regression, even if it isn't everywhere. Although I believe those platforms should fail to add the event which would immediately error also. There just does not appear to be a reason for this to need a sleep, since the EV_ADD should either succeed or fail immediately

**Comment by panjf2000** [comment]
> Yes, that would be still be a regression, even if it isn't everywhere. Although I believe those platforms should fail to add the event which would immediately error also. There just does not appear to be a reason for this to need a sleep, since the EV_ADD alone should either succeed or fail immediately

Okay then,although I'm not certain if the kevent call will return immediately with the wakeup event that was added before the polling.

I'll verify it later and get back to you.

**Comment by panjf2000** [comment]
> If you pass 0 for changes, the documentation says it is guaranteed to return immediately either with success or EINVAL that the filter type is invalid to add

I've verified it and it did return in no time. This sentence in the man pages should endorse this:

> To effect a poll, the timeout argument should be non-NULL, pointing to a zero-valued timespec structure.

**Comment by panjf2000** [comment]
Perfect! That disturbing failure of the flaky test disappeared this time, I think we're good to go!

**Review by saghul — APPROVED** [review]

File: src/unix/async.c
Comment (vtjnash) [reply src/unix/async.c:60]: I am still a bit confused by the complexity here, though it does seem correct now. The things you are checking for `ev[0].filter` is guaranteed to be `EVFILT_USER` here and `ev[0].ident` is guaranteed to be `UV__KQUEUE_EVFILT_USER_IDENT` on every version of kqueue (either with `EV_ERROR` set on an unsupported kernel or not set if it worked). Anything else appearing here would be a pretty significant kernel bug (or userspace stack corruption).

File: src/unix/internal.h
Comment (vtjnash) [reply src/unix/internal.h:516]: Okay, I see that now. It is not really needed anywhere, but it is used in once place. Should we move it into async.c, or just leave it here?

**Review by vtjnash — APPROVED** [review]

**Comment by saghul** [comment]
Cheers! 🥂

**Pull request merged**
